File: XALDAPP/parser/src/main/java/ParseoSms.kt
package com.proyecto.xald.parser

/**
 * Componente interno del contexto de Ingesta.
 * Aplica expresiones regulares al SMS crudo del sistema operativo.
 */
internal class ParseoSms : ServicioParser {

    /**
     * Limpia saltos de línea y normaliza el texto crudo del SMS.
     */
    private fun procesarMensajeCrudo(mensajeCrudo: String): String {
        return mensajeCrudo
            .replace(Regex("[\\r\\n]+"), " ")
            .trim()
    }

    override fun parsear(mensajeCrudo: String): TransaccionProcesadaDTO {
        val textoLimpio = procesarMensajeCrudo(mensajeCrudo)

        val regexMonto = Regex("\\\$([0-9]+(?:\\.[0-9]{1,2})?)")
        val regexComercio = Regex("en\\s+([A-ZÁÉÍÓÚÑa-záéíóúñ\\s]+)")

        val matchMonto = regexMonto.find(textoLimpio)
        val matchComercio = regexComercio.find(textoLimpio)

        val monto = matchMonto?.groupValues?.get(1)?.toDoubleOrNull() ?: 0.0
        val comercio = matchComercio?.groupValues?.get(1)?.trim() ?: "DESCONOCIDO"

        return TransaccionProcesadaDTO(
            monto = monto,
            comercio = comercio,
            textoProcesado = textoLimpio
        )
    }
}

File: XALDAPP_FLU/packages/parser/lib/src/parseo_sms.dart
import 'dtos/transaccion_procesada_dto.dart';
import 'servicio_parser.dart';

class ParseoSms implements ServicioParser {
  // Reglas de expresiones regulares para parseo de SMS bancarios
  static final RegExp _regexMonto = RegExp(r'\$([\d,]+(?:\.\d{2})?)');
  static final RegExp _regexComercio = RegExp(r'en\s+([A-Z0-9\s]+?)(?=\s+por|\s+el|\s+\$|\.)', caseSensitive: false,);

  @override
  TransaccionProcesadaDTO? procesarTextoMensaje(String smsTexto) {
    if (smsTexto.isEmpty) return null;

    final matchMonto = _regexMonto.firstMatch(smsTexto);
    final matchComercio = _regexComercio.firstMatch(smsTexto);

    if (matchMonto == null) return null;

    final montoTexto = matchMonto.group(1)?.replaceAll(',', '') ?? '0';
    final monto = double.tryParse(montoTexto) ?? 0.0;
    final comercio = matchComercio?.group(1)?.trim() ?? 'Comercio Desconocido';

    return TransaccionProcesadaDTO(
      monto: monto,
      comercio: comercio,
      fecha: DateTime.now(),
      smsOriginal: smsTexto,
    );
  }
}